### DISCUSSION TITLE
HUB75 RGB LED matrix with MicroPython on ESP32 NodeMCU?

### DISCUSSION THREAD

**Discussion by danielmader** [micropython/micropython#10776] [General]
Hello,
I'd like to create a simple clock with a 64x32 LED matrix. I've found very valuable information on two websites about how to use this kind of display:
[https://www.bigmessowires.com/2018/05/24/64-x-32-led-matrix-programming/](https://www.bigmessowires.com/2018/05/24/64-x-32-led-matrix-programming/)
[https://www.sparkfun.com/news/2650](https://www.sparkfun.com/news/2650)

My idea was to create the required "images" before and select them with some kind of look-up. Unfortunately, the loops are way too slow. 
Is it impossible to run such a display with MicroPython? Do I need to use CircuitPython instead, with its rgbmatrix module?
```
while True:
    for row in range(16):
        for col in range(64):
            if col < 21:
                R1.value(1)
                G1.value(0)
                B1.value(0)
                R2.value(0)
                G2.value(0)
                B2.value(1)
            elif col < 41:
                R1.value(0)
                G1.value(1)
                B1.value(0)
                R2.value(1)
                G2.value(0)
                B2.value(0)
            else:
                R1.value(0)
                G1.value(0)
                B1.value(1)
                R2.value(0)
                G2.value(1)
                B2.value(0)
            CLK.value(1)  # a) CLK hi to shift in the color bit
            CLK.value(0)  # b) CLK lo to shift in the color bit
        OE.value(1)   # c) disable output
        LAT.value(1)  # d) LAT hi to load the shift register contents into the LED output
        LAT.value(0)  # e) LAT lo to load the shift register contents into the LED output
        A.value(row % 2)
        B.value((row // 2) % 2)
        C.value((row // 4) % 2)
        D.value((row // 8) % 2)
        E.value((row // 16) % 2)
        OE.value(0)   # f) re-enable output
```

Attached is my full code, maybe someone can point out a way to remove the bottleneck. Thanks in advance!
[example_HUB75_Waveshare_64x32_LED.py.txt](https://github.com/micropython/micropython/files/10770533/example_HUB75_Waveshare_64x32_LED.py.txt)

**Comment by russhughes** [comment]
Here is one possible solution: https://notes.benjemmett.com/archives/114

**Reply by danielmader** [reply]
Thank you so much! This looks exactly like what I was looking for. Code quality is very good for my level of understanding! This is a good source to read and learn. Logo display works immediately, the bouncing not yet fully, but I am confident to work it out.

**Reply by danielmader** [reply]
What I do not understand: how do I connect the miso pin which seems nessessary for SPI or SoftSPI:

`class machine.SoftSPI(baudrate=500000, *, polarity=0, phase=0, bits=8, firstbit=MSB, sck=None, mosi=None, miso=None)`

There is no wire for that in the HUB75 connector, of course. Can I simply assign any free pin?

**Reply by russhughes** [reply]
>> Can I simply assign any free pin?

I believe so.

**Reply by benjohnemmett** [reply]
> Can I simply assign any free pin?

That is correct, there is no data to read on MISO so it doesn't matter what pin it is set to. It is a required field for SoftSPI so it must be set to something.

**Reply by danielmader** [reply]
Thank you all, it works with any free pin assigned.

@benjohnemmett: Are you aware of any bug concerning the last column (N=64)? There seems to be an error somewhere when I set a pixel there: for red and green, entire rows light up, for blue the pixel plus the next row. And there are different kinds of "afterglow".

I'll try to attach graphics with different colored pixels set in the last column to illustrate my findings for red-green-blue, each with 
- [0,63]
- [10,63]
- [15,63]
- afterglow after DisplayData()

**blue 0**
![64x32_Images_b0](https://user-images.githubusercontent.com/10500110/220071214-9a7c56b7-8e3f-489f-8607-dd4764397782.png)
**blue 0+10**
![64x32_Images_b0+10](https://user-images.githubusercontent.com/10500110/220071222-ef41a668-5190-44eb-93a8-5a6339ca615e.png)
**blue 0+10+15**
![64x32_Images_b0+10+15](https://user-images.githubusercontent.com/10500110/220071225-5942c38f-da7b-4ab2-abfd-1ae805e0a7a8.png)
**blue afterglow**
![64x32_Images_b-afterglow](https://user-images.githubusercontent.com/10500110/220071227-fd1d0b5a-a0f8-4234-835c-66cf33890bfd.png)

**green 0**
![64x32_Images_g0](https://user-images.githubusercontent.com/10500110/220071230-54d80d3a-1e65-405c-9630-066600e5bc65.png)
**green 0+10**
![64x32_Images_g0+10](https://user-images.githubusercontent.com/10500110/220071234-e507dbb3-db7d-4864-99e1-2b6f4974c11f.png)
**green 0+10+15**
![64x32_Images_g0+10+15](https://user-images.githubusercontent.com/10500110/220071236-04770172-af05-47bf-a654-7eca0872ec6b.png)
**green afterglow**
![64x32_Images_g-afterglow](https://user-images.githubusercontent.com/10500110/220071238-38007312-f264-4846-a0ed-34d12c10a241.png)

**red 0**
![64x32_Images_r0](https://user-images.githubusercontent.com/10500110/220071240-0f491d3a-73e6-4aa1-a6f2-afc17ab309c7.png)
**red 0+10**
![64x32_Images_r0+10](https://user-images.githubusercontent.com/10500110/220071241-1f0e9a1a-3362-49e9-8bd4-0c4fe0f91494.png)
**red 0+10 + 15**
![64x32_Images_r0+10+15](https://user-images.githubusercontent.com/10500110/220071246-0db6a911-6c50-444d-9eba-fe1d4ed20db3.png)
**red afterglow**
![64x32_Images_r-afterglow](https://user-images.githubusercontent.com/10500110/220071251-481e5b44-e355-406c-8fba-6a1634b80250.png)

**Reply by benjohnemmett** [reply]
That problem doesn't sound familiar, but it has been a while since I've used this library. I'll try to find some time to take a deeper look soon. If you have a logic analyzer, it would be good to grab the signals going out.

**Reply by danielmader** [reply]
I'll try to check this at work (have never worked with one myself, though).
Maybe I'm using wrong pins?
![grafik](https://user-images.githubusercontent.com/10500110/220347068-1686e89f-d6f9-4e71-95d0-b1fc2de7e15a.png)

**Reply by benjohnemmett** [reply]
This pinout is different from what I used, which is the default values in Hub75SpiConfiguration. As long as you are setting the pin values in Hub75SpiConfiguration to match your setup before passing it into Hub75Spi then you should be good. Also, there are some nuances about certain ESP-32 pins. You might try swapping around some pins to see if that resolves the issue. I would start with the LAT or OE pins. 

It could definitely also be a bug in my code that I just haven't found before. This library hasn't been used extensively. Now that I look at it again, I'd like to do more thorough tests on it if I can find the time.

**Reply by danielmader** [reply]
I've started with your main.py and changed the pinout there:

```
import hub75
import time
import sys 

config = hub75.Hub75SpiConfiguration()

#config.IlluminationTimeMicroseconds = 25
config.SpiMisoPinNumber = 2  #13 # => nc!!

## row select pins
config.LineSelectAPinNumber = 19  #5
config.LineSelectBPinNumber = 18
config.LineSelectCPinNumber = 5   #19
config.LineSelectDPinNumber = 17  #21
config.LineSelectEPinNumber = 32  #12
## data pins
config.Red1PinNumber   = 25  #2
config.Blue1PinNumber  = 27  #15
config.Green1PinNumber = 26  #4
config.Red2PinNumber   = 14  #16
config.Blue2PinNumber  = 13  #27
config.Green2PinNumber = 12  #17
## clock
config.ClockPinNumber = 16  #22
config.LatchPinNumber = 22  #4   #26
config.OutputEnablePinNumber = 21  #15  #25 # active low
matrix = hub75.Hub75Spi(config)

matrix.ClearDirtyBytes()
coords = [
    # [0,0],
    # [0,62],
    # [31,0,],
    # [31,62],
    [0,63],
    [10,63],
    [15,63],
    [16,63],
    [25,63],
    [31,63],
    # [0,62],
    # [15,62],
    # [16,62],
    # [31,62],
    # [32,62],
    # [31,64]
]

for coord in coords:
    print(">> Is out of bounds?", coord, matrix.IsOutOfBounds(*coord))
    matrix.SetPixelValue(*coord, 4)  # red
    print(">> DirtyBytesArray: ", matrix.DirtyBytesArray)            
    for _ in range(50):
        matrix.DisplayData()
for _ in range(100):
    matrix.DisplayData()
time.sleep(1)

matrix.ClearDirtyBytes()
for coord in coords:
    print(">> Is out of bounds?", coord, matrix.IsOutOfBounds(*coord))
    matrix.SetPixelValue(*coord, 2)  # green
    print(">> DirtyBytesArray: ", matrix.DirtyBytesArray)            
    for _ in range(50):
        matrix.DisplayData()
for _ in range(100):
    matrix.DisplayData()
time.sleep(1)

matrix.ClearDirtyBytes()
for coord in coords:
    print(">> Is out of bounds?", coord, matrix.IsOutOfBounds(*coord))
    matrix.SetPixelValue(*coord, 1)  # blue
    print(">> DirtyBytesArray: ", matrix.DirtyBytesArray)            
    for _ in range(50):
        matrix.DisplayData()
for _ in range(100):
    matrix.DisplayData()
time.sleep(1)
```

**Reply by danielmader** [reply]
Another thing I've noticed: If I use the panel without power, red and green still lights up, but very faint. Actually, this is very desirable for my use case: I am trying to end up with a night clock which must be very faint. Otherwise very low waf (woman acceptance factor) :) I'll be able to turn off power at night using a relay, and turn it back on in the morning for a powerful display during the day. 
This seems odd, though.

**Reply by danielmader** [reply]
I have the same problem with your default pinout from the Hub75 Blaster PCB. The bitbanging approach works no problem (but too slow).

**Reply by benjohnemmett** [reply]
I'm guessing the dim colors are from limited power draw when running with the same power source as the ESP32. Blue LEDs typically require more current, so it would make sense that red & green light up but not blue. If you want dimmer lights with normal operation, you can control it somewhat by reducing the _IlluminationTimeMicroseconds_ value. Also, using some kind of light diffuser would help it work better for use as a bedside clock.

Since you are seeing the same problem with a different pinout, then it's looking like a bug in the code. I'm still pretty swamped right now, but if I get some spare time I'll try to replicate the issue with my setup.

**Reply by benjohnemmett** [reply]
I broke out my logic analyzer today and traced down the issue. The updates are up on the main branch. 

I discovered that the SoftSPI library leaves the MOSI pin at the level of the last bit transferred out. So, if the last bit (right-most pixel) of a given line is high then that pin would remain high while transferring the data for other two color channels causing the entire line to light up. There may be a setting in the SoftSPI library to fix this, however I just manually reset the pins low after shifting data out.

Last pixel illuminated before fix. (Note that the Red1 pin stays high for too long)
![Pasted Graphic](https://user-images.githubusercontent.com/6497121/221336308-b181d816-b2e0-4c54-915e-983ea21ecd29.png)

Last pixel illuminated after fix. (Note that the Red1 pin goes back to LOW before shifting green & blue data)
![Pasted Graphic 2](https://user-images.githubusercontent.com/6497121/221337077-dd70b825-20dd-46e3-8de4-6bc31768dd77.png)

Thanks @danielmader for pointing out the issue and helping me track it down.

**Reply by danielmader** [reply]
This is great news! Thank you very much for your effort!

**Reply by danielmader** [reply]
@all: Ben was very helpful and further optimized his library. I find it very useful for simple Python-only applications that do not require high frame rates or a high amount of coroutine computation.

I've uploaded my code for a simple night clock which is based on it. It only requires an ESP32 module, a 64x32 HUB75 RGB-LED matrix and a power supply. Clock time accuracy is very good and maintained with a timer, and every 12hrs the time is synchronized via NTP:

https://github.com/danielmader/MatrixClock_MicroPythonESP

**Comment by mcauser** [comment]
Have you seen the ESP32-Trinity dev board, which plugs into the back of one of those HUB75 panels?
https://esp32trinity.com/

**Reply by danielmader** [reply]
Haven't seen it yet, but it looks interesting, definitely! Thanks!
However, my bigger problem is not the interfacing but to master the protocoll. Bitbanging is way to slow, and SPI is new to me. So there's a learning curve :)

**Reply by mcauser** [reply]
https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA
Using I2S / DMA will give you fast refresh rates.

**Reply by danielmader** [reply]
True, but this is not a "native" MicroPython solution. There are quite many libs available, it seems, but only very few which allow me to stay with Python.

**Comment by rsautomationembedded** [comment]
Actually P6 rgb module is purchasing, but at this time different type of pixel arragment, so what to do changes into the arduino_gfx library, because my data is not getting proper display

**Reply by benjohnemmett** [reply]
Do you know what protocol your LED matrix uses? I believe that P6 is probably referring to a 6mm space between each LED, not a protocol.
